mmap/src/lib.rs

```diff
@@ -0,0 +1,387 @@
+// SPDX-License-Identifier: Apache-2.0
+
+#![deny(clippy::all)]
+
+use std::io::Result;
+use std::marker::PhantomData;
+use std::ops::*;
+use std::os::raw::c_int;
+use std::os::unix::io::AsRawFd;
+
+use addr::{Address, Offset};
+use bitflags::bitflags;
+use span::Span;
+
+pub use libc::off_t;
+
+// rust-libc is not exhaustive on these flags.
+// For details on the missing ones:
+// https://github.com/enarx/enarx/issues/239
+bitflags! {
+    pub struct Flags: c_int {
+        const ANON = libc::MAP_ANON;
+        const ANONYMOUS = libc::MAP_ANONYMOUS;
+        const BIT32 = libc::MAP_32BIT;
+        const DENYWRITE = libc::MAP_DENYWRITE;
+        const EXECUTABLE = libc::MAP_EXECUTABLE;
+        const FIXED = libc::MAP_FIXED;
+        const FIXED_NOREPLACE = Flags::SYNC.bits << 1;
+        const GROWSDOWN = libc::MAP_GROWSDOWN;
+        const HUGETLB = libc::MAP_HUGETLB;
+        const LOCKED = libc::MAP_LOCKED;
+        const NONBLOCK = libc::MAP_NONBLOCK;
+        const NORESERVE = libc::MAP_NORESERVE;
+        const POPULATE = libc::MAP_POPULATE;
+        const PRIVATE = libc::MAP_PRIVATE;
+        const SHARED = libc::MAP_SHARED;
+        const SHARED_VALIDATE = libc::MAP_SHARED | libc::MAP_PRIVATE;
+        const STACK = libc::MAP_STACK;
+        const SYNC = Flags::HUGETLB.bits << 1;
+    }
+}
+
+bitflags! {
+    pub struct Protections: c_int {
+        const EXEC = libc::PROT_EXEC;
+        const READ = libc::PROT_READ;
+        const WRITE = libc::PROT_WRITE;
+    }
+}
+
+#[derive(Clone)]
+struct Info {
+    span: Span<Address<usize>, Offset<usize>>,
+    protections: Protections,
+    flags: Flags,
+}
+
+pub struct Builder<'a, T> {
+    info: Info,
+    fd: c_int,
+    offset: off_t,
```

Comment:
The `mmap()` C API combines two different primitives: memory offset (expressed by `size_t`) and file offset (expressed by `off_t`). These need to be different because you might be doing 64-bit file offsets on a 32-bit CPU.